hack/builder/Dockerfile

```diff
@@ -47,6 +47,12 @@ RUN dnf install -y dnf-plugins-core && \
     dnf install -y --setopt=install_weak_deps=False \
         gcc-s390x-linux-gnu \
         gcc-c++-s390x-linux-gnu && \
+    dnf groupinstall -y --installroot /usr/s390x-linux-gnu/sys-root --forcearch s390x --disablerepo=epel --releasever 9 \
+        Core && \
+    dnf install -y  --installroot /usr/s390x-linux-gnu/sys-root --forcearch s390x --disablerepo=epel --enablerepo=crb --releasever 9 \
+        glibc-static \
+        libstdc++-static \
+        glibc-devel && \
```

Comment:
IIUC this means that the builder image will always include all dependencies needed to cross-build for s390x.

Considering that the image is already fairly big, and that most people will likely not want to cross-compile on their machines, I'm not convinced this is the best approach.

What happens for aarch64 is that the toolchain is downloaded through Bazel as necessary, so it doesn't need to be included in the container image. That's going to save a lot of disk/bandwidth for the vast majority of users.

That said, I like the idea of getting stuff from EPEL, which implies roughly the same components as those uses for native builds, as opposed to a static toolchain provided by a third party.

So I wonder if we could combine both approaches and get bazeldnf to fetch the RPMs necessary to build a working cross toolchain from EPEL. @zhlhahaha @rmohr what do you think?

Reply:
Hello Andrea, I completely agree with your point. We utilize this Dockerfile for creating multi-architecture images. However, if we retain this code, it will also include s390x cross-compiling packages when building the arm64 builder image, which is not the desired outcome.

In my experience, native builds are the most common use case, making the multi-architecture builder image a necessary component. Cross-compilation, on the other hand, is currently primarily utilized in the release process of kubevirtci. Given that cross-compilation is relatively uncommon, it does make sense to download the cross-compile toolchain when needed, rather than including it by default.

Reply:
> So I wonder if we could combine both approaches and get bazeldnf to fetch the RPMs necessary to build a working cross toolchain from EPEL. @zhlhahaha @rmohr what do you think?

I agree that it would be nicer to handle the toolchain completely through bazel. I mainly choose this approach because i'm new to both bazel and setting up a gcc cross compile enviroment, so it was easier and less complex for me to do it this way. I think it should be possible with bazel + bazeldnf to create the same enviroment, but i don't really know how. If that is the desired way i would need to look it up.

> In my experience, native builds are the most common use case, making the multi-architecture builder image a necessary component. Cross-compilation, on the other hand, is currently primarily utilized in the release process of kubevirtci. Given that cross-compilation is relatively uncommon, it does make sense to download the cross-compile toolchain when needed, rather than including it by default.

I agree that cross-compiling is uncommon. However in the case of s390x, upstream bazel is not supporting s390x (nor wanting to), so i see 3 options for building kubevirt for s390x:
1. Building bazel as a midstream build (does work)
2. Compiling natively without bazel on s390x
3. Cross-compiling on x86

I am right to assume that number 3 is the default go to for building multi-arch releases in the CI? With testing done on native clusters?

Reply:
> In my experience, native builds are the most common use case, making the multi-architecture builder image a necessary component.

Native compilation should also be the simplest one to handle, at least from a Dockerfile point of view: ideally, no changes at all are needed for s390x. There's nothing aarch64-specific right now, for example.

> Cross-compilation, on the other hand, is currently primarily utilized in the release process of kubevirtci.

Which makes cross-compilation something that we absolutely need if we want this merged.

> I agree that it would be nicer to handle the toolchain completely through bazel. I mainly choose this approach because i'm new to both bazel and setting up a gcc cross compile enviroment, so it was easier and less complex for me to do it this way. I think it should be possible with bazel + bazeldnf to create the same enviroment, but i don't really know how. If that is the desired way i would need to look it up.

From a high-level point of view, I think it shouldn't be that much different from what we already do for e.g. launcherbase: tell bazeldnf that we need certain packages, and convert their contents to a tar archive. Then we can define a bazel toolchain in terms of that archive, similarly to what we do today for the aarch64 toolchain downloaded from the Arm website.

Of course the devil is in the detail, and I imagine there are going to be plenty of gotchas along the way :) @zhlhahaha has quite a bit of experience working in this area, so perhaps they could help? Having the same toolchain handling for both aarch64 and s390x would be extremely beneficial IMO.

> I am right to assume that number 3 is the default go to for building multi-arch releases in the CI? With testing done on native clusters?

I believe so, but don't take my word for it.

It's very unfortunate that the bazel project apparently doesn't care about supporting s390x natively. For the purpose of eventually making an official s390x release, cross-compilation should be good enough.

Reply:
Have we agreed on cross-compilation?
Another possibility for not always pulling the s390 cross-build package, would be to create a new image that as parent image the `hack/builder/Dockerfile`. In this way who wants to compile s390x simply pull an additional layer with the dependencies but this won't affect all kubevirt users. The drawback is that we need ot introduce a new image

Reply:
> 1. Building bazel as a midstream build (does work)

I really avoid this. Building bazel from source code takes ages and if there is compilation issue, then it is on us

Reply:
Having the cross-build toolchain as an additional layer would work, but I'm not sure how complicated it would be to integrate it smoothly into the build system. Moreover, if you wanted to make a full build of KubeVirt across architectures, you'd have to juggle three separate images... I'm not sure that would be much of an improvement. The way we're doing cross-compilation for aarch64 right now works really well, the only difference is that in that case there is someone already producing a prepackaged toolchain for us, while for s390x we're trying to assemble one from RPMs. We know that CentOS Stream contains all the necessary packages, so it's "just" a matter of convincing bazel :)